---
title: "Single_Cell_Visum"
author: "Riley Grindle"
date: "2024-04-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
library(Matrix)
library(viridis)
library(tidyverse)
library(Seurat)
library(SeuratData)
library(SeuratObject)
library(SeuratWrappers)
library(Seurat.utils)
library(SingleCellExperiment)
library(gprofiler2)
library(ggplot2)
library(ggsankey)
library(DoubletFinder)
library(stringr)
library(patchwork)
library(loupeR)
library(presto)
library(DropletUtils)
library(PCAtools)
library(aKNNO)
library(jsonlite)
library("rstudioapi")
library(stringi)

options(future.globals.maxSize = 1e9)
options(Seurat.object.assay.version = "v5")
```

```{r}
#####################################################################
# FILL OUT **** parameters.json **** BEFORE EXECUTION OF THE SCRIPT #
#####################################################################
path <- gregexpr(pattern ='/',getSourceEditorContext()$path)
knitr::opts_knit$set(root.dir = (substr(getSourceEditorContext()$path, 1, as.vector(path[[1]])[-1])))

tryCatch({
  json_data <- fromJSON("./parameters.json")
},
error = function(err) {
  message(paste0("The file: ", input, " does not seem to exist, or is not in json format."))
  message("Here's the original error message:\n")
  message(conditionMessage(err))
}
)
```

# Required Parameters

```{r}
# Working Directory
params.workingdirs        <- json_data$working_dir

# Output files directory
params.outdir            <- json_data$out_dir

# 10X Data Input
params.H5_directory      <- json_data$H5_directory
params.h5_name           <- json_data$H5_Name

#   List of Sample names (same order as paths) (Names in single quotes) 
#   eg. c('Sample1','Sample2','Sample3')
params.SampleNames       <- json_data$SampleNames
#   Name of project
params.ProjectName       <- json_data$Project_Name       

# Table To Remove unwanted genes
params.FeatureFile       <- json_data$FeatureFile # Custom feature file containing: MSTR \t MSTR::AMEX::GENE \t Gene Expression
params.Orig_Space_features <- json_data$OrigFeatFile  # Original feature file containing:  MSTR \t Annotation \t Gene Expression
params.ID2Gene           <- json_data$ID2GeneFile
params.gene_subset       <- json_data$SubsetGenes

# Subsetting params
params.subset_qc_metrics        <- json_data$SubsetQCmetrics
params.ncount_subset_quantile   <- json_data$ncountQuantile
params.nfeature_subset_quantile <- json_data$nfeatureQuantile
params.maxMitoPct               <- json_data$maxMitoPct

# Gene Lists for Cell Cycle Scoring and MT
#   Currently set to search for pattern. change to a list "c()" for a custom list
params.organism_ls       <- c("MOUSE", "HUMAN")
params.organism          <- toupper(json_data$Organism)
params.mtGeneList        <- json_data$Mito_Genes


# Seurat Scaling
#    in vars to regress, it should be metadata sections to be regressed or genes i.e. eGFP
params.vars2Regress     <- json_data$Vars2Regress


params.cluster_prefix <- json_data$ClusterPrefix

# Post Scale Options
#    True of False: make Loupe File
params.makeLoupe            <- json_data$MakeLoupe
#    List of Resolutions in ascending order ex. c(0.05,0.1,0.3,0.5,0.7,0.9)
params.resolutions          <- json_data$Resolutions 
#    True of False: Run integration
params.integration          <- json_data$Integrating
#    If params.integration set to true, decide which integration method to use. 
#    (options: CCA, RPCA, Harmony, FastMNN) scVIIntegration needs additional setup -- conda-env
params.integrationMethod   <- json_data$IntegrationMethod                          

# Save Final Merged RDS file at end of Script (Good for keeping Progressive Versions of
# Seurat object)
params.saveRdsSO         <- json_data$SaveRDS

```

#### Set working directory if not in a project

```{r}
if(params.workingdirs[1] != ""){
    setwd(params.workingdirs[1])
}
```

#### Make output directories

```{r}
setwd(params.outdir)

if(!dir.exists("figures")){
    dir.create("figures")
}
if(!dir.exists("markers")){
    dir.create("markers")
}
```

#### Make output directories

```{r}
setwd(params.outdir)

if(!dir.exists("figures/single_cell")){
    dir.create("figures/single_cell")
}
if(!dir.exists("markers/single_cell")){
    dir.create("markers/single_cell")
}
```

#### Removal of unwanted genes

```{r}
setwd(params.workingdirs[1])

if (params.organism == "AXOLOTL"){

ID2Gene <- read.delim(params.ID2Gene,sep = "\t" ,header = F)
rownames(ID2Gene) = ID2Gene[,1]

unannotated.transcripts = rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("PEG10",ID2Gene$V2)] ]
unannotated.transcripts = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("L1TD1",ID2Gene$V2)] ], unannotated.transcripts))
unannotated.transcripts = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("RTL",ID2Gene$V2)] ], unannotated.transcripts))
unannotated.transcripts = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("GIN1",ID2Gene$V2)] ], unannotated.transcripts))
unannotated.transcripts = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("L1\\-RT",ID2Gene$V2)] ], unannotated.transcripts))
unannotated.transcripts = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("^N\\/A",ID2Gene$V2)] ], unannotated.transcripts))
unannotated.transcripts = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("^AMEX",ID2Gene$V2)] ], unannotated.transcripts))

ID2Gene = ID2Gene[!rownames(ID2Gene) %in% unannotated.transcripts,]

rp.genes = rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("^RPL",ID2Gene$V2)]]
rp.genes = unique(c(rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("^RPS",ID2Gene$V2)] ], rp.genes))

ID2Gene = ID2Gene[!rownames(ID2Gene) %in% rp.genes,]

Features <- read.delim(params.FeatureFile,sep = "\t",header = F)
Features[,2][which(duplicated(Features[,2]))]
rownames(Features) = Features[,2]

Features <- Features[Features$V1 %in% rownames(ID2Gene),]

}
```

### Prep Seurat Objects

#### Read in 10X Filtered Directories 

```{r}
setwd(params.workingdirs)

count <- 1
Vector10X <- c()
for (dir in params.workingdirs){
setwd(dir)
Name10X <- paste0(params.SampleNames[count],"_10X")
tryCatch({
  assign(Name10X, Read10X_h5(filename = paste0(params.H5_directory, params.h5_name), use.names = TRUE, unique.features = TRUE))

},error = function(err) {
  message(paste0("\n\nThe file: ",paste0(params.H5_directory, params.h5_name) , " does not seem to exist. Trying again with \"/\" "))
  message("Here's the original error message:\n")
  message(conditionMessage(err))
}, finally = assign(Name10X, Read10X_h5(filename = paste0(params.H5_directory, "/", params.h5_name), use.names = TRUE, unique.features = TRUE))
)
Vector10X <- append(Vector10X, Name10X)
count <- count + 1
}

```

#### Create Seurat Objects from 10X Objects 

```{r}
count <- 1
SeuratObjList <- c()
for (samp in Vector10X){
    NameSO <- paste0("SO_",params.SampleNames[count])
    assign(NameSO, CreateSeuratObject(counts = get(samp), project = params.SampleNames[count], min.cells = 0, min.features = 0))
    SeuratObjList <- append(SeuratObjList, NameSO)
    count <- count + 1
}
```



# Removing unwanted genes - GENE SUBSETTING or GENE Name Joining

```{r}
setwd(params.workingdirs[1])

for (samp in SeuratObjList){
NameSO <- samp
  
if (params.organism == "AXOLOTL"){
SO <- get(NameSO)

# Custom Feature File
features <- read.delim(params.FeatureFile,sep = "\t",header = F)

# Original Spaceranger feature file output
inFeatures <- read.delim(params.Orig_Space_features, sep = "\t", header = F)

inFeatures$ordering <- row.names(inFeatures)

inFeatures$ordering <- str_pad(inFeatures$order, 6, pad = "0" )

outFeatures <- merge(features,inFeatures,by.x = 1,by.y = 1, all = T)

outFeatures <- outFeatures[order(outFeatures$ordering),]

SO.rn <- RenameGenesSeurat(SO,newnames = outFeatures$V2.x, assay = "RNA")

assign(NameSO, SO.rn)

if (params.gene_subset){
  assign(NameSO, get(NameSO)[which(rownames(get(NameSO)) %in% rownames(Features)),])
} else {
}
}
}
```


# Editing out "_" characters in gene names 

```{r}
for (SO in SeuratObjList){
NameSO <- SO
new_features <- rownames(get(SO))

for (i in 1:length(new_features)) {
  
  gene <- new_features[i]
  
  gl <- unlist(strsplit(gene, ""))
  
  if ("_" %in% gl) {
    gene <- gsub("_", "-", gene)
  }
  
  new_features[i] <- gene
  
}


SO.rn <- RenameGenesSeurat(get(NameSO),newnames = new_features, assay = "RNA")
assign(NameSO, SO.rn)

}
```



#### Validate Seurat Objects

```{r}
for (samp in SeuratObjList){
    print(samp)
    print(get(samp))

}
```

#### Cleanup

```{r}
rm(list = Vector10X)
```

## Preprocess each Seurat object individually to remove doublets

#### Select Mitochondrial Genes

```{r}
for (samp in SeuratObjList){

  
if (params.organism == "MOUSE"){
  mito.genes <- row.names(get(NameSO))[grepl(pattern = "^mt-", x = row.names(get(NameSO)))]
  
  calcMT <- function(SO) {
    mito.contig.sampSpec <- intersect(mito.genes , rownames(SO))
    SO[["percent.mt"]] <- PercentageFeatureSet(SO, features = mito.contig.sampSpec)
    SO
}
}else if (params.organism == "HUMAN") {
  mito.genes <- row.names(get(NameSO))[grepl(pattern = "^MT-", x = row.names(get(NameSO)))]
  
  calcMT <- function(SO) {
    mito.contig.sampSpec <- intersect(mito.genes , rownames(SO))
    SO[["percent.mt"]] <- PercentageFeatureSet(SO, features = mito.contig.sampSpec)
    SO
}
}else if (params.organism == "AXOLOTL") {
  mito.genes <- params.mtGeneList
  mito.contig <- c(rownames(ID2Gene)[ID2Gene$V2 %in% mito.genes ], rownames(ID2Gene)[ID2Gene$V2 %in% ID2Gene$V2[grep("^COX",ID2Gene$V2)]] )
  mito.contig <- Features[Features$V1 %in% mito.contig,]
  mito.contig$V2 <- gsub("_","-",mito.contig$V2)
  
  calcMT <- function(SO) {
    mito.contig.sampSpec <- intersect(mito.contig$V2 , rownames(SO))
    SO[["percent.mt"]] <- PercentageFeatureSet(SO, features = mito.contig.sampSpec)
    SO
}
}


print(paste0("Calculating MT% on Sample: ", NameSO))
assign(samp, calcMT(get(samp)))
}
```

#### Normalize Data

```{r}
for (samp in SeuratObjList){
    print(paste0("Normalizing data on Sample: ", samp))
    assign(samp, SCTransform(get(samp), return.only.var.genes = FALSE, assay = 'RNA', vars.to.regress = params.vars2Regress))
}

if (length(SeuratObjList) > 1 ){
count <- 2
countMax <- length(SeuratObjList)
SOlist2up <- c()
while (count <= countMax){
    SOlist2up <- append(SOlist2up, get(SeuratObjList[count]))
    count <- count + 1
}

MergedSO <- merge(get(SeuratObjList[1]), y= SOlist2up, add.cell.ids = params.SampleNames, project = params.ProjectName
)
MergedSO
} else {
  MergedSO <- get(SeuratObjList[1])
  MergedSO
}
```

#Plot QC metrics

```{r}
setwd(params.outdir)

for (samp in SeuratObjList){
    
    print(paste0("Building PrescaleQC plots on sample: ", samp))
    plot1 <- FeatureScatter(get(samp), feature1 = "nCount_RNA", feature2 = "percent.mt", shuffle = T)
    plot2 <- FeatureScatter(get(samp), feature1 = "nCount_RNA", feature2 = "nFeature_RNA", shuffle = T)
    vplot1 <- VlnPlot(get(samp), features = c("nCount_RNA","nFeature_RNA", 'percent.mt'),pt.size = -1)


    pdf(file = paste0('figures/single_cell/',params.ProjectName,"_",samp,'_preFilterQC.pdf'), title = paste0(params.ProjectName," ",samp,' PreFilter QC'), width = 11, height = 8.5)
    print(plot1 + plot2)
    print(vplot1)
    dev.off()
    
    print(plot1 + plot2)
    print(vplot1)

    
  }
```

#### Calculate Subsetting Bounds and Subset

##### Calculate Values

Be sure to validate these values

```{r}
if (params.subset_qc_metrics){

for (samp in SeuratObjList){
  print(paste0("Calcualting subset bounds on sample: ", samp))
  params.minNCount         <- quantile(get(samp)$nCount_RNA, probs = params.ncount_subset_quantile)[[paste0((params.ncount_subset_quantile *100), "%")]]
  params.minNFeature         <- quantile(get(samp)$nFeature_RNA, probs = params.nfeature_subset_quantile)[[paste0((params.nfeature_subset_quantile *100), "%")]]
  

  print(paste0("     Minimum N Count set point:        ", params.minNCount))
  print(paste0("     Minimum N Feature set point:      ", params.minNFeature))
  print(paste0("     Maximum Mito Percent set point:   ", params.maxMitoPct))
  
}

for (samp in SeuratObjList){
    print(paste0("Applying subset bounds on sample: ", samp))
    params.minNCount         <- quantile(get(samp)$nCount_RNA, probs = params.ncount_subset_quantile)[[paste0((params.ncount_subset_quantile *100), "%")]]
    params.minNFeature         <- quantile(get(samp)$nFeature_RNA, probs = params.nfeature_subset_quantile)[[paste0((params.nfeature_subset_quantile *100), "%")]]

    assign(samp,subset(get(samp), subset = nFeature_RNA > params.minNFeature & nCount_RNA > params.minNCount & percent.mt < params.maxMitoPct))
}

for (samp in SeuratObjList){
    print(paste0("Building PrescaleQC plots on sample: ", samp))
    plot1 <- FeatureScatter(get(samp), feature1 = "nCount_RNA", feature2 = "percent.mt", shuffle = T)
    plot2 <- FeatureScatter(get(samp), feature1 = "nCount_RNA", feature2 = "nFeature_RNA", shuffle = T)
    vplot1 <- VlnPlot(get(samp), features = c("nCount_RNA","nFeature_RNA","percent.mt"),pt.size = -1)


    pdf(file = paste0('figures/single_cell/',params.ProjectName,"_",samp,'_postFilterQC.pdf'), title = paste0(params.ProjectName," ",samp,' PostFilter QC'), width = 11, height = 8.5)
    print(plot1 + plot2)
    print(vplot1)
    dev.off()
    
    print(plot1 + plot2)
    print(vplot1)
}

}

```


#### Run PCA on individual samples

```{r}

print("Running PCA on Sample ")
all.genes <- rownames(MergedSO)
MergedSO <- RunPCA(MergedSO, assay = "SCT",features = all.genes ,npcs = 100)

```

#### Generate Elbow Plots and Calculate PCs to use

Validate PCs selected

```{r}
setwd(params.outdir)

params.pcMax <- c()

print(paste0("Generating Elbow Plot for Sample "))
Elbow <- ElbowPlot(MergedSO, ndims = 100, reduction = "pca")
pdf(paste0("./figures/single_cell/",params.ProjectName,"_",paste(params.SampleNames, collapse = "_"),"_ElbowPlot.pdf"), 
    width = 20, height = 15)
print(Elbow)
dev.off()
print(Elbow)
pcCount <- 1
while(Elbow$data$stdev[pcCount]-Elbow$data$stdev[pcCount+1]>0.01 | Elbow$data$stdev[pcCount+1]-Elbow$data$stdev[pcCount+2]>0.01){
    pcCount <- pcCount + 1
}
print(paste0("pc Max Selected at: ", pcCount))
params.pcMax                <- append(params.pcMax,pcCount) 


```

#### Create Cluster names for unintegrated

```{r}
count <- 1
countMax <- length(params.resolutions)
params.resolutionsNames <- c()
while (count <= countMax){
    resVal <- paste0(params.cluster_prefix,params.resolutions[count])
    params.resolutionsNames <- append(params.resolutionsNames, resVal)
    count <- count + 1
}
params.resolutionsNames
```

#### Find Neighbors

```{r}
MergedSO <- FindNeighbors_aKNNO(MergedSO, dims = 1:params.pcMax, reduction = "pca")
```

#### Find Cluster and Set Idents

```{r}
MergedSO <- FindClusters(MergedSO, resolution = params.resolutions, cluster.name = params.resolutionsNames, graph.name="aKNN_O")
```

#### Integrate if `params.integration == T`

```{r}
if (params.integration == T) {
    if(params.integrationMethod == "FastMNN"){
        MergedSO <- IntegrateLayers(object = MergedSO, method = paste0(params.integrationMethod,"Integration"), new.reduction = paste0("integrated.",params.integrationMethod))
    }else{
        MergedSO <- IntegrateLayers(object = MergedSO, method = paste0(params.integrationMethod,"Integration"), orig.reduction = "pca", new.reduction = paste0("integrated.",params.integrationMethod))
    }
}
```

#### Create Cluster Names for Integrated

```{r}
if (params.integration == T) {
count <- 1
countMax <- length(params.resolutions)
params.resolutionsNames <- c()
while (count <= countMax){
    resVal <- paste0(params.integrationMethod,params.cluster_prefix,params.resolutions[count])
    params.resolutionsNames <- append(params.resolutionsNames, resVal)
    count <- count + 1
}
params.resolutionsNames
}
```

#### Find Neighbors for Integrated

```{r}
if (params.integration == T) {
MergedSO <- FindNeighbors_aKNNO(MergedSO, dims = 1:params.pcMax, reduction = paste0("integrated.",params.integrationMethod))
}
```

#### Find Clusters and Set Ident for Integrated

```{r}
if (params.integration == T) {
MergedSO <- FindClusters(MergedSO, resolution = params.resolutions, cluster.name = params.resolutionsNames, graph.name="aKNN_O")
}
```




# Generate Sankey Plot - SC

```{r}
setwd(params.outdir)

key <- T
integrated <- F
while (key) {
  
mergedSOmetaRes <- MergedSO@meta.data

if (params.integration == F){
    mergedSOmetaRes <- mergedSOmetaRes[,grepl(paste0("^", params.cluster_prefix),names(mergedSOmetaRes))]
    names(mergedSOmetaRes) <- gsub(params.cluster_prefix,"res",names(mergedSOmetaRes))
    integrated <- F
    key <- F
    plot <- "Unintegrated"
}else{
    pattern <- paste0(params.integrationMethod,"Res")
    mergedSOmetaRes <- mergedSOmetaRes[,grepl(pattern,names(mergedSOmetaRes))]
    names(mergedSOmetaRes) <- gsub(pattern,"res",names(mergedSOmetaRes))
    integrated <- T
    params.integration <- F
    plot <- "Integrated"
}



df <- mergedSOmetaRes%>%
    make_long(names(mergedSOmetaRes))

df$next_node <- str_pad(df$next_node,2, pad = "0")
df$node <- str_pad(df$node,2, pad = "0")

df$reswClust <- paste0(df$x,": ", df$node)

reagg <- df%>%
    dplyr::group_by(reswClust)%>%                                        # Get counts 
    tally()

df2 <- merge(df, 
             reagg, 
             by.x = 'reswClust', 
             by.y = 'reswClust', 
             all.x = TRUE)

assign(plot, ggplot(df2, aes(x = x,                        
                      next_x = next_x,                                     
                      node = node,
                      next_node = next_node,        
                      fill = factor(node),
                      label = paste0(node, " = ", n))) +                 # Creates lable with node and count
    geom_sankey(flow.alpha = 0.5,                                        # Add opacity
                node.color = "black",                                    # Node Color
                show.legend = TRUE) +                                    # This determines if you want your legend to show
    geom_sankey_label(size = 3,
                      color = "black", 
                      fill = "white") +                                  # This specifies the Label format for each node 
    theme_bw() + 
    theme(legend.position = 'none') + 
    theme(axis.title = element_blank(),
          axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          panel.grid = element_blank()) + 
    scale_fill_viridis_d(option = "H") + 
    labs(title = paste0("Sankey: ",params.ProjectName, ifelse(integrated ==T,paste0(" Integrated (", params.integrationMethod,")"), " Unintegrated")," Dataset")) + 
    labs(caption =NULL ) + 
    labs(fill = 'Nodes'))



### Save Sankey Plot
pdf(paste0("./figures/single_cell/",params.ProjectName,ifelse(integrated ==T,paste0("Integrated__", params.integrationMethod), "_Unintegrated"),"SankeyPlot.pdf"), width = 20, height = 15)
print(get(plot))
}
dev.off()

MergedSO[["RNA"]] <- JoinLayers(MergedSO[["RNA"]])
params.integration <- json_data$Integrating

```


#### Find Clusters and Generate Markers

```{r}
setwd(params.outdir)

if (length(SeuratObjList) > 1){
  MergedSO <- PrepSCTFindMarkers(MergedSO, assay = "SCT", verbose = TRUE)
}

marker_ls <- c()

if(params.integration ==T){
  for(i in 1:length(params.resolutions)){
    print(paste0("Finding Markers at ",params.resolutions[i]," resolution..."))
    Idents(MergedSO) <- paste0(params.integrationMethod,"Res.",params.resolutions[i])
    assign(paste0("MarkersRes_Integrated",params.resolutions[i]),FindAllMarkers(MergedSO, only.pos = T, logfc.threshold = 0.3))
    marker_ls <- append(marker_ls, paste0("MarkersRes_Integrated",params.resolutions[i]))
  }
}  
  
### Find Clusters and Generated Markers
for(i in 1:length(params.resolutions)){
    print(paste0("Finding Markers at ",params.resolutions[i]," resolution..."))
    Idents(MergedSO) <- paste0(params.cluster_prefix,params.resolutions[i])
    assign(paste0("MarkersRes",params.resolutions[i]),FindAllMarkers(MergedSO, only.pos = T, logfc.threshold = 0.3))
    marker_ls <- append(marker_ls, paste0("MarkersRes",params.resolutions[i]))
}
setwd(params.outdir)
# Add in annotation info and output tsv files
for(j in 1:length(marker_ls)){
    if (length(levels(factor(get(marker_ls[j])$cluster))) <= 1) {
      next
    }
    print(paste0("Generating Markers tsv at ",marker_ls[j]," resolution..."))
    markertemp <- get(paste0(marker_ls[j]))
    markertemp <- markertemp  %>% arrange(cluster, desc(avg_log2FC))
    markertemp$markercall <- 1
    markertemp <- markertemp %>%
        mutate(markercall = ifelse(avg_log2FC >= 1.7 & markertemp$pct.1/markertemp$pct.2 >= 1.7 & -log10(markertemp$p_val_adj) >= 27, 1, 0))
    assign(marker_ls[j],markertemp)
    write.table(get(marker_ls[j]), file = paste0("./markers/single_cell/",marker_ls[j],".tsv"), sep = "\t", quote = F,row.names = F)
    rm(markertemp)
}
```

#### Set Default Ident

```{r}
params.identRes             <-   0.2   # Set resolution eg. 0.3 
    
### Set Ident
if(params.integration ==T ){
    Idents(MergedSO) <-paste0(params.integrationMethod,"Res.",params.identRes)
}else{
    Idents(MergedSO) <-paste0(params.cluster_prefix,params.identRes)
}
```

#### Run UMAP and TSNE

```{r}
MergedSO <- RunUMAP(MergedSO, dims = 1:params.pcMax, reduction = "pca", reduction.name = "umap.unintegrated")
if(params.integration ==T ){
    MergedSO <- RunUMAP(MergedSO, dims = 1:params.pcMax, reduction = paste0("integrated.",params.integrationMethod), reduction.name = paste0("umap.",params.integrationMethod))
}
MergedSO <- RunTSNE(MergedSO, dims = 1:params.pcMax, reduction = "pca", reduction.name = "tsne.unintegrated")
if(params.integration ==T ){
    MergedSO <- RunTSNE(MergedSO, dims = 1:params.pcMax, reduction = paste0("integrated.",params.integrationMethod), reduction.name = paste0("tsne.",params.integrationMethod))
}


```

#### Plot inital UMAPs

```{r}
setwd(params.outdir)

pdf(paste0("./figures/single_cell/",params.ProjectName,"InitialUMAPs.pdf"), width = 20, height = 15)
if(params.integration ==T ){
    DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T) +
        DimPlot(object = MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = .25, label = T, group.by = "ident", shuffle = T)
}else{
    DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T) 
}
dev.off()

if(params.integration ==T ){
    print(DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T) +
        DimPlot(object = MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = .25, label = T, group.by = "ident", shuffle = T))
}else{
    print(DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'umap.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T)) 
}
```

#### Plot initial TSNE

```{r}
setwd(params.outdir)

pdf(paste0("./figures/single_cell/",params.ProjectName,"InitialTSNEs.pdf"), width = 20, height = 15)
if(params.integration ==T ){
    DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T) +
        DimPlot(object = MergedSO, reduction = paste0("tsne.",params.integrationMethod), pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = paste0("tsne.",params.integrationMethod), pt.size = .25, label = T, group.by = "ident", shuffle = T)
}else{
    DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T) 
}
dev.off()

if(params.integration ==T ){
    print(DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T) +
        DimPlot(object = MergedSO, reduction = paste0("tsne.",params.integrationMethod), pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = paste0("tsne.",params.integrationMethod), pt.size = .25, label = T, group.by = "ident", shuffle = T))
}else{
    print(DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "orig.ident", shuffle = T) + 
        DimPlot(object = MergedSO, reduction = 'tsne.unintegrated', pt.size = .25, label = T, group.by = "ident", shuffle = T)) 
}
```

#### QC Plots on UMAP


```{r}
setwd(params.outdir)

pdf(paste0("./figures/single_cell/",params.ProjectName,"qcUMAPs.pdf"), width = 20, height = 15)
if(params.integration ==T ){
    FeaturePlot(MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = 1, features = "nFeature_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = 1, features = "nCount_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = 1, features = "percent.mt", order = T, cols = c("lightgrey", "blue")) + 
        plot_layout(ncol = 2, nrow = 2)
}else{
    FeaturePlot(MergedSO, reduction = 'umap.unintegrated', pt.size = 1, features = "nFeature_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = 'umap.unintegrated', pt.size = 1, features = "nCount_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = 'umap.unintegrated', pt.size = 1, features = "percent.mt", order = T, cols = c("lightgrey", "blue")) + 
        plot_layout(ncol = 2, nrow = 2)
}
dev.off()

if(params.integration ==T ){
    print(FeaturePlot(MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = 1, features = "nFeature_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = 1, features = "nCount_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = paste0("umap.",params.integrationMethod), pt.size = 1, features = "percent.mt", order = T, cols = c("lightgrey", "blue")) + 
        plot_layout(ncol = 2, nrow = 2))
}else{
    print(FeaturePlot(MergedSO, reduction = 'umap.unintegrated', pt.size = 1, features = "nFeature_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = 'umap.unintegrated', pt.size = 1, features = "nCount_RNA", order = T, cols = c("lightgrey", "blue")) +
        FeaturePlot(MergedSO, reduction = 'umap.unintegrated', pt.size = 1, features = "percent.mt", order = T, cols = c("lightgrey", "blue")) + 
        plot_layout(ncol = 2, nrow = 2))
}

```

#### Make Loupe File

```{r}
setwd(params.outdir)

if(params.makeLoupe == T){
    if(!dir.exists("LoupeFiles")){
        dir.create("LoupeFiles")
    }
    create_loupe(count_mat = MergedSO@assays$RNA$counts,
                 clusters = select_clusters(MergedSO),
                 projections = select_projections(MergedSO),
                 output_name = paste0("./LoupeFiles/",params.ProjectName), 
                 force = T
    )
}
```

#### Save RDS

```{r}
setwd(params.outdir)

if(params.saveRdsSO == T){
    if(!dir.exists("RDSObjects")){
        dir.create("RDSObjects")
    }
    if (!dir.exists("RDSObjects/single_cell")){
        dir.create("RDSObjects/single_cell")
    }
    saveRDS(MergedSO, paste0("./RDSObjects/single_cell/",params.ProjectName,"SO.rds"))
}
```


# Generate Cluster Meta Data .csv to project on Loupe File


```{r}
setwd(params.outdir)

resolutions = colnames(MergedSO@meta.data)[grepl(pattern = params.cluster_prefix, colnames(MergedSO@meta.data))]

cell_cluster_idents <- FetchData(MergedSO, vars = resolutions)
umap <- MergedSO@reductions$umap.unintegrated@cell.embeddings
tsne <- MergedSO@reductions$tsne.unintegrated@cell.embeddings

if (params.integration){
  umap.i <- MergedSO@reductions$umap.Harmony@cell.embeddings
  tsne.i <- MergedSO@reductions$tsne.Harmony@cell.embeddings
}

if (!dir.exists("projections")) {
  dir.create("projections")
}

slice_id <- function(string, count) {
  tmp_str <- stri_reverse(string)
  underscore_index <- gregexpr("_", tmp_str)[[1]]
  print(tmp_str)
  sliced_string <- substr(tmp_str, start = 1, stop = underscore_index-1)
  sliced_string <- stri_reverse(sliced_string)
  sliced_string <- substr(sliced_string, start = 1, stop = nchar(sliced_string)-2)
  sliced_string <- paste0(sliced_string, "-", count)
  return(sliced_string)
}


if (length(SeuratObjList) > 1){
  counter <- 1
  combined_clust <- as.data.frame(matrix(nrow = 0, ncol = ncol(cell_cluster_idents)))
  combined_umap <- as.data.frame(matrix(nrow = 0, ncol = ncol(cell_cluster_idents)))
  combined_tsne <- as.data.frame(matrix(nrow = 0, ncol = ncol(cell_cluster_idents)))
  combined_umap.i <- as.data.frame(matrix(nrow = 0, ncol = ncol(cell_cluster_idents)))
  combined_tsne.i <- as.data.frame(matrix(nrow = 0, ncol = ncol(cell_cluster_idents)))
  for (name in params.SampleNames) {
    rows_clust <- which(grepl(name, rownames(cell_cluster_idents)))
    tmp_clust <- cell_cluster_idents[rows_clust,]
    rownames(tmp_clust) <- lapply(rownames(tmp_clust), slice_id, count=counter)
    rows_umap <- which(grepl(name, rownames(umap)))
    tmp_umap <- umap[rows_umap,]
    rownames(tmp_umap) <- lapply(rownames(tmp_umap), slice_id, count=counter)
    rows_tsne <- which(grepl(name, rownames(tsne)))
    tmp_tsne <- tsne[rows_tsne,]
    rownames(tmp_tsne) <- lapply(rownames(tmp_tsne), slice_id, count=counter)
    combined_clust <- rbind(combined_clust, tmp_clust)
    combined_umap <- rbind(combined_umap, tmp_umap)
    combined_tsne <- rbind(combined_tsne, tmp_tsne)
    if (params.integration) {
      rows_umap.i <- which(grepl(name, rownames(umap.i)))
      tmp_umap.i <- umap.i[rows_umap.i,]
      rownames(tmp_umap.i) <- lapply(rownames(tmp_umap.i), slice_id, count=counter)
      rows_tsne.i <- which(grepl(name, rownames(tsne.i)))
      tmp_tsne.i <- tsne.i[rows_tsne.i,]
      rownames(tmp_tsne.i) <- lapply(rownames(tmp_tsne.i), slice_id, count=counter)
      combined_umap.i <- rbind(combined_umap.i, tmp_umap.i)
      combined_tsne.i <- rbind(combined_tsne.i, tmp_tsne.i)
    }
    counter <- counter + 1
  }
  write.csv(combined_clust, file = "projections/seurat_clusters.csv", col.names = NA, quote = FALSE, row.names = TRUE)
  write.csv(combined_umap, file = paste0("projections/seurat_umap.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
  write.csv(combined_tsne, file = paste0("projections/seurat_tsne.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
  if (params.integration){
    write.csv(combined_umap.i, file = paste0("projections/seurat_umap.integrated.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
  write.csv(combined_tsne.i, file = paste0("projections/seurat_tsne.integrated.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
  }
  
}else {
write.csv(umap, file = paste0("projections/seurat_umap.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
write.csv(tsne, file = paste0("projections/seurat_tsne.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
write.csv(cell_cluster_idents, file = paste0("projections/seurat_clusters.", params.ProjectName,".csv"), col.names = NA, quote = FALSE, row.names = TRUE)
}
```



# Confusion matrix between 2 similar Seurat Object experiments


```{r}
setwd(params.outdir)

library(caret)

params.file_1_title      <- "Only Annotated Features" #Axis title for naming experiment in plot (ex: subsetted axolotl limb)
params.file_2_title      <- "All Features" #Axis title for naming experiment in plot (ex: Full exp axolotl limb)
params.comparison_title  <- "feature_subset" #Title for comparison being made (ex: feature_subset)


file_1 <- "RDSObjects/single_cell/pmurawala_001_neo_spatialSO_filtered.rds"
file_2 <- "RDSObjects/single_cell/pmurawala_001_neo_spatialSO_unfiltered.rds"

SO_1 <- LoadSeuratRds(file_1)
SO_2 <- LoadSeuratRds(file_2)

col_fraction <- function(cell, position, df){
  
  tmp_vector <- c()
  for (i in 1:length(cell)){
    fraction <- (cell[i]/sum(df[,position[i]]))
    tmp_vector <- append(tmp_vector, fraction)
  }
  return(tmp_vector)
  
}

if (!dir.exists("./figures/confusion_matrices/")) {
  dir.create("./figures/confusion_matrices/")
}

for (i in 1:ncol(SO_1@meta.data)){
  df1 <- SO_1@meta.data
  df2 <- SO_2@meta.data
  colname <- colnames(df1[i])
  if (grepl("Res", colname)){
    filtered <- factor(df1[,i])
    unfiltered   <- factor(df2[,i])
    
    if (length(levels(filtered)) <= length(levels(unfiltered))){
      Data <- filtered
      Ref <- unfiltered
      params.y_title <- params.file_1_title
      params.x_title <- params.file_2_title
      params.flip <- F
    } else {
      Data <- unfiltered
      Ref <- filtered
      params.y_title <- params.file_2_title
      params.x_title <- params.file_1_title
      params.flip <- T
    }
    
    example <- confusionMatrix(data=Data, reference=Ref)
    
    scaled <- as.data.frame(matrix(ncol = ncol(example$table), nrow = 0))
    for (i in 1:nrow(example$table)){
      row = example$table[i,]
      row_fract <- (row/sum(row))
  
      indices <- seq(length(row))
  
      col_fract <- col_fraction(row, indices, example$table)
    
      sum_fract <- (row_fract + col_fract)
  
      scaled <- rbind(scaled, sum_fract)
    }
    colnames(scaled) <- seq(0,ncol(scaled)-1)
    rownames(scaled) <- seq(0,nrow(scaled)-1)
    
    data <- melt(t(scaled))
    
    data[,1] <- paste0("Cluster_",data[,1])
    data[,2] <- paste0("Cluster_",data[,2])
    
    labels_x <- unique(data[,1])
    labels_y <- unique(data[,2])
    
    data$Var1 <- factor(data$Var1, levels = labels_x)
    data$Var2 <- factor(data$Var2, levels = labels_y)
    colnames(data) <- c("Var1", "Var2", "Combined_Fraction")
    
    pdf(paste0("./figures/confusion_matrices/",params.ProjectName,".", colname,".", params.comparison_title, ".confusion_plot.pdf"), width = 30, height = 25)
    heat_map <- ggplot(data = data, aes(x = Var1, y = Var2)) +
      geom_tile(aes(fill = Combined_Fraction), color = "white") +
      scale_fill_gradient(low = "lightblue", high = "darkblue", guide = guide_colorbar(barwidth = 5, barheight = 20,
                                                                                       theme = theme(legend.text = element_text(size = 20)))) +
      geom_text(aes(label = round(Combined_Fraction, 3)), color = "white" , size = 15) +
      labs(x = params.x_title, y = params.y_title, fill = "Combined_Fraction", title = colname) +
      theme_minimal()
    if (params.flip){
      heat_map <- heat_map + coord_flip()
    } else {
      
    }
    
    heat_map <- heat_map + theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 26),
            axis.text.y = element_text(hjust = 1, vjust = 1, size = 26),
            axis.title.x = element_text(face = "bold", vjust = 0.25, size = 40, margin = margin(t = 50)),
            axis.title.y = element_text(face = "bold", vjust = 0.25, size = 40, margin = margin(r = 50)) ,
            title = element_text(face = "bold", size = 48, margin = margin(t = 50)),
            legend.title=element_text(size=25, vjust = 7))
    
    print(heat_map)
    dev.off()
    print(heat_map)

  }
}
```
















